File: face_recognition/actor_scraper/actor_scraper.py
#
# Actor Scraper
# Scrapes actor images, creates json file of actor info
# Author: Phillip T.
#

###########
# OPTIONS #
###########

base_url = 'https://www.imdb.com/name/'
nconsts_file = 'nconsts/nconsts.json'
json_dest = 'actors.json'
img_dest = 'actors'

################################################################################

import urllib.request
import requests
import json

from bs4 import BeautifulSoup
from requests import get
from PIL import Image

# set up actors list
actors = []

# load nconsts file
with open(nconsts_file) as json_file:
  nconsts = json.load(json_file)

# remove duplicates
nconsts = list(set(nconsts))

# find length of nconsts list
nconsts_length = len(nconsts)

# loop through nconsts
for index, nconst in enumerate(nconsts):
  # construct actor URL, filename, and image path
  actor_url = base_url + nconst
  filename = f'{nconst}.jpg'
  img_path = f'{img_dest}/{filename}'
  
  # obtain response from page
  response = get(actor_url)
  
  # parse HTML text
  html = BeautifulSoup(response.text, 'html.parser')
  
  # find actor div
  actor_div = html.find('div', {'id': 'name-overview-widget'})
  
  # extract actor name and image URL
  actor_name = actor_div.h1.span.text
  img_url = actor_div.find('img', {'id': 'name-poster'})['src']
  
  # create actor dictonary and add to actors to list
  actor = { 'nconst': nconst,
            'name': actor_name,
            'filename': filename }
  actors.append(actor)

  # download and store image
  r = requests.get(img_url, stream = True)
  i = Image.open(r.raw)
  i.save(img_path)
  
  # notify user
  print(f'[PROG] {nconst} - {index + 1} of {nconsts_length} actors processed')

# write actor info to json file
with open(json_dest, 'w') as outfile:
  json.dump(actors, outfile)

print('[DONE] All actors processed')

File: face_recognition/actor_scraper/actor_scraper.js
/*
*
* Actor Scraper
* Extracts actor info from top 1000 imdb page
* Author: Phillip T.
*
*/

/////////////
// OPTIONS //
/////////////

const baseUrl = 'https://www.imdb.com/list/ls058011111/'
const totalPages = 10
const actorsPerPage = 100
const actorsFile = 'actors/actors.json'

////////////////////////////////////////////////////////////////////////////////

const rp = require('request-promise')
const $ = require('cheerio')
const fs = require('fs')

let actors = []
let operationsCompleted = 0;
let totalActors = totalPages * actorsPerPage;

const main = () => {
  console.log("[INIT] scraping actors...")
  getActors()
}

// get actor info
const getActors = () => {
  // loop through all pages of the list
  for (let pageNumber = 1; pageNumber <= totalPages; pageNumber++) {
    // format url for extration
    let url = `${baseUrl}?page=${pageNumber}`
    
    // extract actor info for each page
    rp(url)
    .then(function(html){
      for (let i = 0; i < actorsPerPage; i++) {
        // create actor object
        let actor = {}

        // extract nconst, add to object
        actor['nconst'] = $('.lister-item-image > a', html)[i].attribs.href
        .replace('/name/', '').replace('/?ref_=nmls_pst', '')

        // extract name, add to object
        actor['name'] = $('.lister-item-header a', html).eq(i).text().trim()

        // push object to array
        actors.push(actor)

        // increment operations counter
        operation()
      }

      // display progress to user
      console.log(`[PROG] ${actors.length} of ${totalActors} processed`)
    })
  }
}

// save actors to JSON file
const saveActors = () => {
  // convert actors to json
  const json = JSON.stringify(actors)

  // write to file
  fs.writeFileSync(actorsFile, json)

  // notify user
  console.log('[DONE] actors written to file', actorsFile)
  console.log(actors.length)
}

// when getNconsts done, run SaveNconsts
// hacky way not to use async or promises: https://stackoverflow.com/a/27914617
const operation = () => {
  // increment operations counter
  operationsCompleted++

  // once all actors are scraped, run saveActors
  if (operationsCompleted === totalActors) saveActors(actors)
}

main()